containers/daap-python-base/src/var/task/data_product_metadata.py

```diff
@@ -1,116 +1,291 @@
+                "Parameters": {"field.delim": ",", "escape.delim": "\\"},
+            },
+            "BucketColumns": [],
+            "SortColumns": [],
+            "Parameters": {},
+            "StoredAsSubDirectories": False,
+        },
+        "PartitionKeys": [],
+        "TableType": "EXTERNAL_TABLE",
+        "Parameters": {"classification": "csv", "skip.header.line.count": "1"},
+    },
+}
 
-def get_data_product_metadata_spec_path(version: str = "") -> str:
-    # if version is empty we'll get the latest version
-    if version == "":
-        file_paths = get_filepaths_from_s3_folder(
-            DataProductConfig.metadata_spec_prefix().uri
-        )
+
+def get_data_product_specification_path(
+    spec_type: JsonSchemaName, version: None | str = None
+) -> str:
+    """gets path for given version or latest version of metadata and schema json schema"""
+
+    # if version is empty we get the latest version
+    if version is None:
+        file_paths = get_filepaths_from_s3_folder(specification_prefix(spec_type).uri)
         versions = list(
             {i for p in file_paths for i in p.split("/") if i.startswith("v")}
         )
         versions.sort(key=lambda x: [int(y.replace("v", "")) for y in x.split(".")])
         latest_version = versions[-1]
-        path = DataProductConfig.metadata_spec_path(latest_version)
+        path = specification_path(spec_type, latest_version)
     else:
-        DataProductConfig.metadata_spec_path(version)
-        path = DataProductConfig.metadata_spec_path(version)
+        path = specification_path(spec_type, version)
 
     return path.uri
 
 
-def split_bucket_and_key(path):
-    bucket, key = path.replace("s3://", "").split("/", 1)
-    return bucket, key
-
-
-class DataProductMetadata:
-    """
-    class to handle creation and updating of
-    metadata relating to a dataproduct
-    """
+class DataProductBaseJsonSchema:
+    """base class for operations on json type metadata and schema for data products"""
 
-    def __init__(self, data_product_name: str, logger: DataPlatformLogger):
+    def __init__(
+        self,
+        data_product_name: str,
+        logger: DataPlatformLogger,
+        json_type: JsonSchemaName,
+        write_bucket_path: BucketPath,
+    ):
         self.data_product_name = data_product_name
         self.logger = logger
-        bucket, key = split_bucket_and_key(
-            DataProductConfig(name=data_product_name).metadata_path().uri
+        self.valid = False
+        self.type = json_type
+        self.write_bucket = write_bucket_path.bucket
+        self.write_key = write_bucket_path.key
+        self.exists = self._check_if_metadata_or_schema_exists(
+            self.write_bucket, self.write_key, self.type
         )
-        self.metadata_bucket = bucket
-        self.metadata_key = key
-        self._check_if_metadata_exists()
 
-    def _check_if_metadata_exists(self) -> object:
+        if not self.exists:
```

Comment:
I think this gives us two sources of truth for the data product version - this logic, and the `write_key`.

I'm concerned we could get a conflict where the table schema is written using 1.1 but it's used metadata from 1.0 or something.

I'm also not sure I follow the logic - if write path is /foo/v1.1/metadata.json, and it doesn't exist, why should it fall back to v1.0?

Maybe we should extract the version from the write key instead of basing it on the existence of the schema/metadata.